src/main/scala/release.scala

```diff
@@ -0,0 +1,193 @@
+package ohnosequences.db.taxonomy
+
+import ohnosequences.s3._
+import ohnosequences.forests.{EmptyTree, NonEmptyTree, Tree}, Tree._
+import java.io.File
+import helpers._
+
+case object release {
+
+  /** Generates Good, Environmental or Unclassified tree from the Full one
+    *
+    * @param fullTree the whole [[TaxTree]]
+    * @param treeType a type from: [[TreeType.Good]], [[TreeType.Environmental]]
+    * or [[TreeType.Unclassified]]
+    */
+  private def generateTree(fullTree: TaxTree, treeType: TreeType): TaxTree =
+    fullTree match {
+      case fullTree: EmptyTree[TaxNode] => fullTree
+      case fullTree: NonEmptyTree[TaxNode] =>
+        treeType match {
+          case TreeType.Full => fullTree
+          case TreeType.Unclassified =>
+            val positions = whichPositions(fullTree) { pos =>
+              TreeType.get(fullTree(pos)) == TreeType.Unclassified
+            }.toSet
+
+            coveringTree(fullTree, positions)
+          case TreeType.Environmental =>
+            val positions = whichPositions(fullTree) { pos =>
+              TreeType.get(fullTree(pos)) == TreeType.Environmental
+            }.toSet
+
+            coveringTree(fullTree, positions)
+          case TreeType.Good =>
+            filter(fullTree) { node =>
+              TreeType.get(node) == TreeType.Good
+            }
```

Comment:
What does filter do? Does it remove any branch that starts with a node that does not fulfill the predicate?

Reply:
Correct, it does exactly what you have said. Note that this is not the same as pruning those branches, since we would prune only the children of the node, not the node itself.